File: DATA_STRUCTURES_PY/Tree_AVL__(Balance_binary_tree).py
"""/*--AVL Tree - Сбалансированное двоичное дерево--*/"""
class Node:
    def __init__(self): pass

class AVL:
    def __init__(self): pass                # Конструктор
    def _height(self, node): pass			# Спектр высоты (Возвращает высоту для текущего поддерева, для пустого = 0)
    def _bfactor(self, node): pass          # Вычисление баланса узла (Возвращает разницу высоты между левым и правым поддеревом)
    def _fixheight(self, node): pass		# Установка поля высоты узла (Обновляет поле height для текущего узла)
    def _rotate_right(self, node): pass		# Правый поворот вокруг p
    def _rotate_left(self, node): pass		# Левый поворот вокруг q
    def _balance(self, node): pass   		# Балансировка узла p
    def _insert(self, node, key): pass	    # Вставка ключа k в дерево с корнем p
    def _find_min(self, node): pass			# Служебная функция поиска минимума для функции удаления узла
    def _remove_min(self, node): pass		# Служебная функция удаления узла с минимальным ключа
    def _remove_value(self, node, key): pass  # Удаление ключа из дерева
    def _print(self, node, level): pass     # Печать структуры дерева
    def _del_tree(self, node): pass         # Рекурсивное освобождение памяти
    def insert(self, key): pass             # Вставка
    def remove(self, key): pass				# Удаление по значению
    def show(self): pass				    # Вывод на экран
"""---------------------------------------------------------------------------"""

class Node:
    def __init__(self, data):
        self._data = data                   # Данные
        self._left = None                   # Левый потомок
        self._right = None                  # Правый потомок
        self._height = 1                    # Высота поддерева


class AVL:
    def __init__(self):                     # Конструктор
        self._root = None

    def _height(self, node):      # Спектр высоты (Возвращает высоту для текущего поддерева, для пустого = 0)
        if node is not None:
            return node._height
        else:
            return 0
        # return node._height if node is not None else None

    def _bfactor(self, node):     # Вычисление баланса узла (Возвращает разницу высоты между левым и правым поддеревом)
        return self._height(node._right) - self._height(node._left)

    def _fixheight(self, node):           # Установка поля высоты узла (Обновляет поле height для текущего узла)
        hl = self._height(node._left)
        hr = self._height(node._right)
        if hl > hr:
            node._height = hl + 1
        else:
            node._height = hr + 1
        #node._height = (hl + 1) if hl > hr else (hr + 1)

    def _rotate_right(self, node):       # Правый поворот вокруг p
        q = node._left
        node._left = q._right
        q._right = node
        self._fixheight(node)
        self._fixheight(q)
        return q

    def _rotate_left(self, node):       # Левый поворот вокруг q
        p = node._right
        node._right = p._left
        p._left = node
        self._fixheight(node)
        self._fixheight(p)
        return p

    def _balance(self, node):           # Балансировка узла p
        self._fixheight(node)
        if self._bfactor(node) == 2:
            if self._bfactor(node._right) < 0:
                node._right = self._rotate_right(node._right)
            return self._rotate_left(node)
        if self._bfactor(node) == -2:
            if self._bfactor(node._left) > 0:
                node._left = self._rotate_left(node._left)
            return self._rotate_right(node)
        return node

    def _insert(self, node, key):            # Вставка ключа k в дерево с корнем p
        if node is None:
            return Node(key)
        if key < node._data:
            node._left = self._insert(node._left, key)
        else:
            node._right = self._insert(node._right, key)
        return self._balance(node)

    def _find_min(self, node):               # Служебная функция поиска минимума для функции удаления узла
        return self._find_min(node._left) if node._left else node

    def _remove_min(self, node):             # Служебная функция удаления узла с минимальным ключа
        if node._left is None:
            return node._right
        node._left = self._remove_min(node._left)
        return self._balance(node)

    def _remove_value(self, node, key):     # Удаление ключа из дерева
        if node is None:
            return None
        if key < node._data:
            node._left = self._remove_value(node._left, key)
        elif key > node._data:
            node._right = self._remove_value(node._right, key)
        else:
            q = node._left
            r = node._right
            node = None
            if r is None:
                return q
            minv = self._find_min(r)
            minv._right = self._remove_min(r)
            minv._left = q
            return self._balance(minv)
        return self._balance(node)

    def _print(self, node, level):        # Печать структуры дерева
        if node is None:
            return
        for i in range(level):
            print("---", end=' ')
        print(node._data)
        self._print(node._left, level + 1)
        self._print(node._right, level + 1)

    def _del_tree(self, node):            # Рекурсивное освобождение памяти
        if node:
            self._del_tree(node._left)
            self._del_tree(node._right)
            node._data = None

    def insert(self, key):                   # Вставка
        self._root = self._insert(self._root, key)

    def remove(self, key):                   # Удаление по значению
        self._remove_value(self._root, key)

    def show(self):                          # Вывод на экран
        self._print(self._root, 0)


def main():
    avl = AVL()
    for i in range(20):
        avl.insert(i)
    avl.show()

if __name__ == "__main__":
    main()

File: DATA_STRUCTURES_CPP/Tree_AVL__(Balance_binary_tree).cpp
/*--AVL Tree - Сбалансированное двоичное дерево--*/
#include <iostream>
using std::cout;

template <typename Type>
class AVL {
private:
	class Node {
	public:
		Type data;						// Поле с данными (ключ)
		int height;						// Высота поддерева (либо 0, либо -1, либо 1)
		Node *left;						// Указатель на левый лист
		Node *right;					// Указатель на правы лист
		// Конструктор с параметрами
		Node(Type dat) : data(dat), height(1), left(nullptr), right(nullptr) {}
	};
	Node  *root;						// Указатель на корень дерева
	int   height(Node *p);				// Спектр высоты (Возвращает высоту для текущего поддерева, для пустого = 0)
	int   bfactor(Node* p);				// Вычисление баланса узла (Возвращает разницу высоты между левым и правым поддеревом)
	void  fixheight(Node* p);			// Установка поля высоты узла (Обновляет поле height для текущего узла)
	Node* rotate_right(Node *p);		// Правый поворот вокруг p
	Node* rotate_left(Node *q);			// Левый поворот вокруг q
	Node* balance(Node *p);				// Балансировка узла p
	Node* insert(Node *p, Type k);		// Вставка ключа к в дерево с корнем p	
	Node* find_min(Node *n);			// Служебная функция поиска минимума для функции удаления узла
	Node* remove_min(Node *n);			// Служебная функция удаления узла с минимальным ключа
	Node* remove_value(Node *n, Type k);// Удаление ключа из дерева
	void  print(Node* p, int level);	// Печать структуры дерева
	void  del_tree(Node *n);			// Для деструктора, рекурсивное освобождение памяти
public:
	AVL(): root(nullptr) {};			// Конструктор
	~AVL();								// Деструктор
	void insert(Type key);				// Вставка
	void remove(Type key);				// Удаление по значению
	void print();						// Вывод на экран
};


template <typename Type>
int AVL<Type>::height(Node* p) {
	return p ? p->height : 0;
}

template <typename Type>
int AVL<Type>::bfactor(Node* p) {
	return height(p->right) - height(p->left);
}

// Запрашиваются выстоы поддеревьев выбирается наибольшая и к ней прибавляется 1
template<typename Type>
void AVL<Type>::fixheight(Node* p) {
	int hl = height(p->left);
	int hr = height(p->right);
	p->height = (hl > hr ? hl : hr) + 1;
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::rotate_right(Node* p) {
	Node *q = p->left;
	p->left = q->right;
	q->right = p;
	fixheight(p);
	fixheight(q);
	return q;
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::rotate_left(Node* q) {
	Node* p = q->right;
	q->right = p->left;
	p->left = q;
	fixheight(q);
	fixheight(p);
	return p;
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::balance(Node* p) {
	fixheight(p);
	if (bfactor(p) == 2) {
		if (bfactor(p->right) < 0)
			p->right = rotate_right(p->right);
		return rotate_left(p);
	}
	if (bfactor(p) == -2) {
		if (bfactor(p->left) > 0)
			p->left = rotate_left(p->left);
		return rotate_right(p);
	}
	return p;
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::insert(Node* p, Type k) {
	if (!p)
		return new Node(k);
	if (k < p->data)
		p->left = insert(p->left, k);
	else
		p->right = insert(p->right, k);
	return balance(p);
}

template<typename Type>
void AVL<Type>::print(Node* p, int level) {
	int i;
	if (p == nullptr)
		return;
	for (i = 0; i < level; ++i)
		cout << "---";
	cout << p->data << "\n";
	print(p->left, level + 1);
	//cout << p->data << "\n";
	print(p->right, level + 1);
	//cout << p->data << "\n";
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::find_min(Node* n) {
	return n->left ? find_min(n->left) : n;
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::remove_min(Node* n) {
	if (n->left == nullptr)
		return n->right;
	n->left = remove_min(n->left);
	return balance(n);
}

template<typename Type>
typename AVL<Type>::Node* AVL<Type>::remove_value(Node* n, Type k) {
	if (!n)
		return nullptr;
	if (k < n->data)
		n->left = remove_value(n->left, k);
	else if (k > n->data)
		n->right = remove_value(n->right, k);
	else {	// k == n->data
		Node *q = n->left;
		Node *r = n->right;
		delete n;
		if (!r)
			return q;
		Node *min = find_min(r);
		min->right = remove_min(r);
		min->left = q;
		return balance(min);
	}
	return balance(n);
}

template<typename Type>
void AVL<Type>::del_tree(Node* n) {
	if (n) {
		del_tree(n->left);
		del_tree(n->right);
		delete n;
	}
}

template<typename Type>
AVL<Type>::~AVL() {
	del_tree(root);
}

template<typename Type>
void AVL<Type>::insert(Type key) {
	root = insert(root, key);
}

template<typename Type>
void AVL<Type>::remove(Type key) {
	remove_value(root, key);
}

template<typename Type>
void AVL<Type>::print() {
	print(root, 0);
}


int main() {
	AVL<int> avl;
	for (int i = 0; i < 10; ++i) {
		avl.insert(i);
		//avl.print();
	}
	avl.print();
	avl.remove(5);
	avl.remove(8);
	avl.remove(1);
	avl.print();
	return 0;
}